app/src/sensors.c

```diff
@@ -18,65 +18,144 @@ LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+    { .triggers_per_rotation = DT_PROP_OR(ZMK_KEYMAP_SENSORS_NODE, triggers_per_rotation, 20) }
 
-    LOG_DBG("sensor %d", item->sensor_number);
+    LISTIFY(ZMK_KEYMAP_SENSORS_LEN, SENSOR_CHILD_DEFAULTS, (, ), 0)
+#endif
+};
+
+static struct sensors_item_cfg sensors[] = {LISTIFY(ZMK_KEYMAP_SENSORS_LEN, SENSOR_ITEM, (, ), 0)};
+
+static K_SEM_DEFINE(pending_sensors_sem, 1, 1);
+static uint32_t pending_sensors;
+
+const struct zmk_sensor_config *zmk_sensors_get_config_at_index(uint8_t sensor_index) {
+    if (sensor_index > ARRAY_SIZE(configs)) {
+        return NULL;
+    }
 
-    err = sensor_sample_fetch(dev);
+    return &configs[sensor_index];
+}
+
+static void trigger_sensor_data_for_position(uint32_t sensor_index) {
+    int err;
+    const struct sensors_item_cfg *item = &sensors[sensor_index];
+
+    err = sensor_sample_fetch(item->dev);
     if (err) {
         LOG_WRN("Failed to fetch sample from device %d", err);
         return;
     }
 
-    ZMK_EVENT_RAISE(new_zmk_sensor_event((struct zmk_sensor_event){
-        .sensor_number = item->sensor_number, .sensor = dev, .timestamp = k_uptime_get()}));
+    struct sensor_value value;
+    err = sensor_channel_get(item->dev, item->trigger.chan, &value);
+
+    if (err) {
+        LOG_WRN("Failed to get channel data from device %d", err);
+        return;
+    }
+
+    ZMK_EVENT_RAISE(new_zmk_sensor_event(
+        (struct zmk_sensor_event){.sensor_index = item->sensor_index,
+                                  .channel_data = {(struct zmk_sensor_channel_data){
+                                      .value = value, .channel = item->trigger.chan}},
+                                  .timestamp = k_uptime_get()}));
 }
 
-static void zmk_sensors_init_item(const char *node, uint8_t i, uint8_t abs_i) {
-    LOG_DBG("Init %s at index %d with sensor_number %d", node, i, abs_i);
+static void run_sensors_data_trigger(struct k_work *work) {
+    if (k_sem_take(&pending_sensors_sem, K_FOREVER) < 0) {
+        return;
+    }
+
+    uint32_t pd = pending_sensors;
+    pending_sensors = 0;
+
+    k_sem_give(&pending_sensors_sem);
+
+    for (int i = 0; pd != 0; i++, pd >>= 1, i++) {
+
+        if ((pd & BIT(0)) != 0) {
+            LOG_DBG("pending data for %d", i);
+            trigger_sensor_data_for_position(i);
+        }
+    }
+}
+
+K_WORK_DEFINE(sensor_data_work, run_sensors_data_trigger);
+
+static void zmk_sensors_trigger_handler(const struct device *dev,
+                                        const struct sensor_trigger *trigger) {
+    const struct sensors_item_cfg *test_item =
+        CONTAINER_OF(trigger, struct sensors_item_cfg, trigger);
+    int sensor_index = test_item - sensors;
+
+    if (sensor_index < 0 || sensor_index >= ARRAY_SIZE(sensors)) {
+        LOG_ERR("Invalid sensor item triggered our callback");
+        return;
+    }
+
+    if (k_is_in_isr()) {
+        if (k_sem_take(&pending_sensors_sem, K_MSEC(1)) >= 0) {
```

Comment:
The documentation says "The kernel does allow an ISR to take a semaphore, however the ISR must not attempt to wait if the semaphore is unavailable." Does that mean that the wait time should be 0? It seems like that could lead to dropped sensor events if run_sensors_data_trigger() was running.

I think you could do this more reliably with an atomic variable instead. This part would just be
```c
atomic_set_bit(&pending_sensors, sensor_index);
```
and you could either change run_sensors_data_trigger to use
```c
pd = atomic_set(&pending_sensors, 0);
for (...)
```
or simplify the loop to
```c
for (int i = 0; i < ARRAY_SIZE(sensors); i++) {
    if (atomic_test_and_clear_bit(&pending_sensors, i)) {
        trigger_sensor_data_for_position(i);
    }
}
```
(this should work just the same, though it isn't _exactly_ the same since if a bit gets set while it's iterating through the loop, it could get immediately handled instead of waiting until the next call to run_sensors_data_trigger.)

Reply:
As a bonus, `ATOMIC_DEFINE(<sensor count>)` would automatically expand to the needed size no matter how many sensors there are instead of silently breaking when you go beyond 32.

Reply:
Switched to the atomic approach, tested on ZMK Uno w/ the QDEC driver that causes the ISR code to be hit, working fine. Thanks!